### DISCUSSION TITLE
installation failure on Mac 'ERROR: Failed building wheel for h5py'

### DISCUSSION THREAD

**Discussion by jkingslake** [firedrakeproject/firedrake#2851] [Firedrake support] [answered]
I am installing on mac 13.2.1 (22D68).

It seems to fail when trying to install h5py.  I tried installing h5py with homebrew separately, but that did not seem to help. 

[This](https://github.com/firedrakeproject/firedrake/issues/1619) issue seems to be the same,  but it is unclear if it was resolved. 

Log files are attached. 

Any help would be much appreciated.

[firedrake-install.log](https://github.com/firedrakeproject/firedrake/files/11133830/firedrake-install.log)
PETSC log files:
[configure.log](https://github.com/firedrakeproject/firedrake/files/11133838/configure.log)
[make.log](https://github.com/firedrakeproject/firedrake/files/11133839/make.log)

**Comment by connorjward** [comment]
I notice two things:
- You are using the system Python. This is [known to cause problems](https://firedrakeproject.org/download.html#system-anti-requirements). Please use a homebrew-installed Python version instead.
- I'm seeing a number of warnings like:
    ```
    ld: warning: dylib (/Users/jkingslake/firedrake/src/petsc/default/lib/libmpi.dylib) was built for newer macOS version (10.15) than being linked (10.9))
    ```
    This surprises me because 10.9 is really quite an old version to be cropping up here. Is your `firedrake-install` freshly downloaded? And is your xcode installation up to date?

**Comment by jkingslake** [comment]
Thanks so much for taking a look so quickly!

## System Python
I was aware of this issue and tried to avoid it, but I didnt find much online about how to determine if I am using the system python. I ran `brew install python` prior to this installation and that all seemed to work. How do I tell if I am using the system python? How do I ensure that I am using the homebrew version for the installation?

## newer macOS version...

I downloaded the install script a few weeks ago. Should i get a more recent one?

I did update xcode while dealing with another issue with this installation. I can try doing that again if you think it will help.

**Reply by jkingslake** [reply]
update: I am trying adding export PATH="/usr/local/opt/python/libexec/bin:$PATH" to .zshrc as suggested [here](https://stackoverflow.com/questions/5157678/how-do-i-use-brew-installed-python-as-the-default-python)

**Comment by jkingslake** [comment, marked as answer]
The installation is now working!   :partying_face:  !

Two things I did to solve this are:

1. At this stage I had already installed python with homebrew using `brew install python`, it just wasnt using this version yet. To correct this I added 
```export PATH="/usr/local/opt/python/libexec/bin:$PATH"```
 to .zshrc, as suggested [here](https://stackoverflow.com/questions/5157678/how-do-i-use-brew-installed-python-as-the-default-python). This ensures that the homebrew-installed version of python is the default python you get when you type `python` into the terminal. I think adding the same line to .bash_profile will work the same way if your terminal is set up to use bash. 
2. I download the latest firedrake installation script using 
```wget https://raw.githubusercontent.com/firedrakeproject/firedrake/master/scripts/firedrake-install```
Then ran it.

In case they are useful to anyone, here are the firedrake and petsc log files from the successful installation: 
[firedrake-install.log](https://github.com/firedrakeproject/firedrake/files/11141881/firedrake-install.log)
[configure.log](https://github.com/firedrakeproject/firedrake/files/11141885/configure.log)
[make.log](https://github.com/firedrakeproject/firedrake/files/11141894/make.log)

Note, prior to this I was having an issue with a fortran compiler, so I followed the advice [here](https://mac.install.guide/commandlinetools/3.html) and reinstalled the xcode command line tools.

Thanks @connorjward for the help!

**Reply by jkingslake** [reply]
I wrote down the steps I followed to install firedrake and [icepack](https://github.com/icepack/icepack) on my 2018 macbook air [here](https://github.com/ldeo-glaciology/LEAP-Cryo-planning/blob/main/icepack_install.md), in case it is useful to anyone. 

@connorjward, I wonder if it would be useful to include some of these steps in the docs. Maybe around [here](https://github.com/firedrakeproject/firedrake/blob/master/docs/source/download.rst#debugging-install-problems). Maybe in the form of FAQs? I would be happy to contribute a PR if it can be useful for the next person trying to tackle this installation.

**Reply by connorjward** [reply]
Thanks @jkingslake. It would be nice to include some of this in the installation instructions. Specifically I think it would be good to mention:
- One should use `$(brew --prefix)/bin/python3` to make sure you're using the right Python. I think this is better than telling users to modify their `PATH`.
- Add some instructions for updating xcode. I know a number of users hit this.
- Maybe it would be useful to mention that users should always try to download a fresh install script as we frequently modify it.

This sort of thing might be good to add to the flowchart we already have: https://firedrakeproject.org/download.html#debugging-install-problems.

**Comment by hwoo749** [comment]
Hiya,

I have successfully downloaded firedrake, and I have been following your instructions @jkingslake on how to install icepack, which have been very helpful - I think I have installed it correctly but I am struggling to activate the virtual environment.
I get this message:

harrietwoodward@Harriets-MacBook-Air ~ % source firedrake/bin/activate
(firedrake) harrietwoodward@Harriets-MacBook-Air ~ % cd $VIRTUAL_ENV/src/icepack
cd: no such file or directory: /Users/harrietwoodward/firedrake/src/icepack

The path to  icepack is as follows:
![Screen Shot 2023-05-10 at 9 34 16 PM](https://github.com/firedrakeproject/firedrake/assets/80790825/32f5b3a8-a3c8-46fc-bf60-8f484655c0a8)


How do I go about activating the icepack virtual environment? I tried cd $VIRTUAL_ENV/firedrake/bin/firedrake/src/icepack
but this gave me the same error message of no such file or directory. Any help would be greatly appreciated!

**Reply by connorjward** [reply]
Your icepack is installed into `firedrake/bin/firedrake/src/icepack`. It should actually be `firedrake/src/icepack`. I think you should be able to simply run `pip install git+https://github.com/icepack/icepack.git#egg=icepack` to have it installed correctly.

**Reply by hwoo749** [reply]
Hi there, thank you for the quick reply. :))

I ran that command, and this installed icepack in this path:
![Screen Shot 2023-05-11 at 11 10 56 AM](https://github.com/firedrakeproject/firedrake/assets/80790825/91929243-6cd8-4d5e-aafe-2812ba1dfcc6)


How do I go about activating the virtual environment from this path?

**Reply by ReubenHill** [reply]
I think you should be able to do it with `$ source /harrietwoodward/firedrake/bin/activate`

**Reply by hwoo749** [reply]
Hiya, I think I've got it sorted! Thank you for your help. :))

I ran the tests and got this message - apparently they should be fine though right? Only warnings, no errors.

===================================== test session starts ======================================
platform darwin -- Python 3.11.3, pytest-7.3.1, pluggy-1.0.0
rootdir: /Users/harrietwoodward/firedrake/bin/firedrake/src/icepack
plugins: nbval-0.10.0, mpi-0.1, xdist-3.2.1, anyio-3.6.2
collecting ... firedrake:WARNING OMP_NUM_THREADS is not set or is set to a value greater than 1, we suggest setting OMP_NUM_THREADS=1 to improve performance
collected 4 items                                                                              

test/ice_shelf_test.py log(error) ~= 1.90382 * log(dx) -28.0289
log(error) ~= 2.91626 * log(dx) -42.3439
log(error) ~= 3.92242 * log(dx) -56.6201
.log(error) ~= 1.90382 * log(dx) -28.0289
log(error) ~= 2.91626 * log(dx) -42.3439
log(error) ~= 3.91985 * log(dx) -56.5973
.log(error) ~= 2.4978 * log(dx) -45.7326
..

======================================= warnings summary =======================================
test/ice_shelf_test.py::test_diagnostic_solver_convergence[icepack]
test/ice_shelf_test.py::test_diagnostic_solver_convergence[icepack]
  /Users/harrietwoodward/firedrake/src/loopy/loopy/target/c/compyte/dtypes.py:184: DeprecationWarning: `np.bool8` is a deprecated alias for `np.bool_`.  (Deprecated NumPy 1.24)
    reg.get_or_register_dtype("bool", np.bool8)

test/ice_shelf_test.py: 12248 warnings
  /Users/harrietwoodward/firedrake/src/ufl/ufl/core/expr.py:276: DeprecationWarning: Expr.ufl_domain() is deprecated, please use extract_unique_domain(expr) instead.
    warnings.warn("Expr.ufl_domain() is deprecated, please "

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
======================== 4 passed, 12250 warnings in 572.51s (0:09:32) =========================

**Reply by ReubenHill** [reply]
You can surpress `firedrake:WARNING OMP_NUM_THREADS is not set or is set to a value greater than 1, we suggest setting OMP_NUM_THREADS=1 to improve performance` by doing what it says: `$ export OMP_NUM_THREADS = 1` (add this to your `bashrc` to avoid needing to do it regularly).

The others you don't need to worry about.
